File: 241001/트로미노/tromino.cpp
#include <iostream>
#include <algorithm>
using namespace std;
#define DIR_NUM 4
#define MAX_NM 200

int n, m, ans;
int grid[MAX_NM][MAX_NM];
int dx[DIR_NUM] = {-1, 0, 1, 0}, dy[DIR_NUM] = {0, 1, 0, -1};

bool InRange(int x, int y) {
    return 0<=x && x<n && 0<=y && y<n;
}

int FindMax(int x, int y) {
    int num = grid[x][y];
    int sum = 0;

    for(int d=0; d<DIR_NUM; d++) {
        int nx1 = x+dx[d], ny1 = y+dy[d];
        int nx2 = x+dx[(d+1)%4], ny2 = y+dy[(d+1)%4];
        if(!InRange(nx1, ny1) || !InRange(nx2, ny2)) continue;

        sum = max(sum, num+grid[nx1][ny1]+grid[nx2][ny2]);
    }

    for(int d=0; d<2; d++) {
        int nx1 = x+dx[d], ny1 = y+dy[d];
        int nx2 = x+dx[(d+2)%4], ny2 = y+dy[(d+2)%4];
        if(!InRange(nx1, ny1) || !InRange(nx2, ny2)) continue;

        sum = max(sum, num+grid[nx1][ny1]+grid[nx2][ny2]);
    }

    return sum;
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n>>m;

    for(int i=0; i<n; i++) {
        for(int j=0; j<m; j++) {
            cin >> grid[i][j];
        }
    }

    for(int i=0; i<n; i++) {
        for(int j=0; j<m; j++) {
            ans = max(ans, FindMax(i, j));
        }
    }

    cout << ans;

    return 0;
}

File: 241104/트로미노/tromino.py
n, m = tuple(map(int, input().split()))

grid = [
    list(map(int, input().split()))
    for _ in range(n)
]

dx, dy = [-1, 0, 1, 0], [0, 1, 0, -1]
ans = 0

def inRange(x, y):
    return 0<=x<n and 0<=y<m

def calc1(x, y):

    result = 0
    for d in range(4):
        nx1, ny1 = x+dx[d], y+dy[d]
        nx2, ny2 = x+dx[(d+1)%4], y+dy[(d+1)%4]

        if inRange(nx1, ny1) and inRange(nx2, ny2):
            result = max(result, grid[x][y]+grid[nx1][ny1]+grid[nx2][ny2])
    
    return result

def calc2(x, y):
    result = 0
    for d in range(2):
        nx1, ny1 = x+dx[d], y+dy[d]
        nx2, ny2 = x+dx[(d+2)%4], y+dy[(d+2)%4]

        if inRange(nx1, ny1) and inRange(nx2, ny2):
            result = max(result, grid[x][y]+grid[nx1][ny1]+grid[nx2][ny2])
    
    return result

for i in range(n):
    for j in range(m):
        ans = max(ans, calc1(i, j))
        ans = max(ans, calc2(i, j))

print(ans)